Add PlayerMixin to use advancements and rotation to detect player motility

donington opened the pull request:
Using advancements to detect various player actions and interactions with the world around them to trigger motile.

Also added basic rotation detection as a starting point, but it is specifically disabled in vehicles since they cause the player to rotate triggering false positives.

Options to disable these features have been added as well.

donington committed:
Add PlayerMixin to use advancements to trigger motility

donington committed:
Add rudimentary rotation detection to PlayerMixin

Rotation detection is specifically disabled in vehicles because it
currently doesn't work properly in that scenario.

bytzo commented on src/main/java/net/bytzo/sessility/SessilityProperties.java:
```diff
@@ -14,6 +14,10 @@ public class SessilityProperties extends Settings<SessilityProperties> {
+	public final boolean advancementActionDetection = this.get("advancement-action-detection", true);
+	public final boolean advancementInteractDetection = this.get("advancement-interact-detection", true);
+	public final boolean advancementDebugMessages = this.get("advancement-debug-messages", false);
+	public final boolean rotationTriggersMotility = this.get("rotation-triggers-motility", true);
```
For consistency, we might want to name properties relating to enabling/disabling things starting with verbs. For example, `detect-advancement-action` and `detect-advancement-interaction` are more consistent with properties like `hide-sessile-in-tab-list` and `skip-sessile-in-sleep-count`.

bytzo commented on src/main/java/net/bytzo/sessility/mixins/PlayerMixin.java:
```diff
@@ -0,0 +1,132 @@
+		Stream.of("crafted", "talked_to_villager", "traded_with_villager", "eat_cake_slice", "fill_cauldron", "use_cauldron",
+							"play_noteblock", "tune_noteblock", "pot_flower", "trigger_trapped_chest", "enchant_item",
+							"play_record", "bell_ring")
+		.collect(Collectors.collectingAndThen(Collectors.toCollection(HashSet::new),Collections::unmodifiableSet));
```
Note that Sessility already marks players as motile when they:

- Start/Abort/Finish breaking a block
- Place a block
- Use item(s)
- Carry item(s) within an inventory
- Send a chat message
- Swing their arm (punching)
- Sneak/Leave a bed/Sprint/Jump a vehicle/Opens a vehicle inventory/Activate their elytra
- Punch an entity
- Send a command
- Set item(s) in a container
- Select a crafting recipe
- Click a menu button (enchanting table, loom, etc.)

So, advancement actions such as `mined`, `used`, `drop`, etc. and advancement interactions like `eat_cake_slice`, `play_noteblock`, `bell_ring`, etc. may be redundant. Advancement actions like `jump` are still very useful though! Perhaps the list of advancement triggers could be as minimal as possible, and only contain actions that are most often missed from the list above (like jumping)?

bytzo commented on src/main/java/net/bytzo/sessility/mixins/PlayerMixin.java:
```diff
@@ -0,0 +1,132 @@
+		.collect(Collectors.collectingAndThen(Collectors.toCollection(HashSet::new),Collections::unmodifiableSet));
+
+	/* optionally prints advancement debugging information */
+	private void advancement_debug(String state, String playername, String advancement) {
```
We might want to annotate this with `@Unique`, just in case another mod injects a similarly named method into the class.

bytzo commented on src/main/java/net/bytzo/sessility/mixins/PlayerMixin.java:
```diff
@@ -0,0 +1,132 @@
+	}
+
+	/* use advancements as potential motility detectors */
+	private void processAdvancement(ServerPlayer player, String advancement, int count) {
```
Similar to the comment above, we might want to annotate this with `@Unique` to avoid potential naming conflicts with other mods injecting into the same class.

bytzo commented on src/main/java/net/bytzo/sessility/mixins/PlayerMixin.java:
```diff
@@ -0,0 +1,132 @@
+		}
+
+		// process advancement interact list and wildcard filters
+		if (Sessility.settings().properties().advancementInteractDetection &&
+				 (advancementInteractList.contains(advancement) ||
+					advancement.startsWith("interact_with_") ||
+					advancement.startsWith("inspect_") ||
+					advancement.startsWith("open_")
+				 )
+			 ) {
```
Nit: There seem to be accidental spaces in the indentation of lines 78 and 83. Additionally, whenever a single statement is broken into multiple, two tabs should be used.

```java
if (
		Sessility.settings().properties().advancementInteractDetection &&
		(
				advancementInteractList.contains(advancement) ||
				advancement.startsWith("interact_with_") ||
				advancement.startsWith("inspect_") ||
				advancement.startsWith("open_")
		)
) {
```

Well okay, I must admit, my style in this case isn't great :P

In situations like this though, I usually just extract parts into variables for readability.

```java
var advancementInteracts =
		advancementInteractList.contains(advancement) ||
		advancement.startsWith("interact_with_") ||
		advancement.startsWith("inspect_") ||
		advancement.startsWith("open_");
if (Sessility.settings().properties().advancementInteractDetection && advancementInteracts) {
```

bytzo commented on src/main/java/net/bytzo/sessility/mixins/PlayerMixin.java:
```diff
@@ -0,0 +1,132 @@
+		// process advancement action list
+		if (Sessility.settings().properties().advancementActionDetection && advancementActionList.contains(advancement)) {
+			advancement_debug("[action]", player.getName().getString(), advancement);
+			player.resetLastActionTime();
```
Instead of calling `resetLastActionTime()`, we might want to call `setSessile(false)` instead. This way, we aren't modifying the behavior of vanilla's idle kick.

donington replied:
I did originally try to do it this way, but because Sessility relies on that internal timer, any call to `setSessile()` does technically work but is immediately outdated by the state of the last action timer.  Calling `resetLastActionTime()` directly was the workaround.

I think it could be overcome if `SessilePlayer`'s state was backed by a separate timestamp variable and a lastActionTime tracker.  In `postTick()` the `idleTime` could then be updated against this timestamp or if the `getLastActionTime()` tracker is stale, and proceed as normal.  Finally, a call to `setSessile(false)` would then be able to just update the timestamp via `Util.getMillis()`, allowing other sources than the `resetLastActionTime()` to override the sessile state.

bytzo replied:
Ah, you're right! We'd need to maintain our own last action time field to make this work.

bytzo commented on src/main/java/net/bytzo/sessility/mixins/PlayerMixin.java:
```diff
@@ -0,0 +1,132 @@
+				 )
+			 ) {
+			advancement_debug("[interact]", player.getName().getString(), advancement);
+			player.resetLastActionTime();
```
Similar to the above, we might not want to call `resetLastActionTime()` at it will affect the behavior of the vanilla server idle kick. Instead, we could use `setSessile(false)`.

bytzo commented on src/main/java/net/bytzo/sessility/mixins/PlayerMixin.java:
```diff
@@ -0,0 +1,132 @@
+		// process advancement interact list and wildcard filters
+		if (Sessility.settings().properties().advancementInteractDetection &&
+				 (advancementInteractList.contains(advancement) ||
+					advancement.startsWith("interact_with_") ||
+					advancement.startsWith("inspect_") ||
+					advancement.startsWith("open_")
```
As noted earlier, Sessility already marks players as motile when they:

- Start/Abort/Finish breaking a block
- Place a block
- Use item(s)
- Carry item(s) within an inventory
- Send a chat message
- Swing their arm (punching)
- Sneak/Leave a bed/Sprint/Jump a vehicle/Opens a vehicle inventory/Activate their elytra
- Punch an entity
- Send a command
- Set item(s) in a container
- Select a crafting recipe
- Click a menu button (enchanting table, loom, etc.)

So, filtering advancements that start with `interact_with_`, `inspect_`, and `open_` might be redundant. However, I haven't looked closely at all the possible advancements. Are there some that don't overlap with the list above?

donington replied:
I hadn't had time to finish processing them all after sorting to see if there were any edge cases.  You're correct though, a lot of this is absolutely redundant.  Probably just going to remove anything related to the interaction filter since it appears all the cases are already covered by your system!

bytzo commented on src/main/java/net/bytzo/sessility/mixins/PlayerMixin.java:
```diff
@@ -0,0 +1,132 @@
+
+		// detect player rotation
+		if (last_xRot != xRot || last_yRot != yRot)
+			player.resetLastActionTime();
```
Similar to the above, we might want to avoid calling `resetLastActionTime()` at could affect the vanilla server idle kick. Instead, you could use `setSessile(false)`.

donington replied:
This and the other code was written from a perspective of detecting motility, and not sessility.  So in the case as above, it would actually be attempting to `setSessile(false)`.

bytzo reviewed (changes requested):
Thanks for working on this contribution so far! :D

Currently, I am wondering which advancements we should use by default. Since Sessility already handles most cases where a player interacts with the world, I think it could be sufficient to subscribe to only the actions that are most missed, such as jumping. This way, the list can be kept as minimal as possible.

Also, thanks for adding the rotation tracker to this! That is something I've always wanted to add, but I never got around to it.

donington committed:
remove ServerPlayerMixin's dependancy on getLastActionTime()

  - added lastMotileTime and actionTimeTracker to allow calls to
    setSessile(false) to update state without altering the server
    lastActionTime state, but allowing sessility to be updated

donington committed:
Code cleanup and optimization

  - Revised variable names for config file.
  - Removed redundant action advancements that are already covered
    by Sessility (just jump remains).
  - Removed all advancement interaction stuff and configuration since
    it was entirely redundant (Sessility covers all those cases already).
  - Removed advancement debug messages.
  - Replaced all direct calls to ServerPlayer.resetLastActionTime()
    with SessilePlayer.setSessile(false).
  - remove unused imports from PlayerMixin
  - optimize lastMotileTime in ServerPlayerMixin
  - convert spaces to tabs and comment formatting cleanup

donington commented:
Updated.  I did attempt at adding the support needed for `setSessile(false)` operate correctly.  Let me know if I missed anything!

bytzo commented on src/main/java/net/bytzo/sessility/mixins/ServerPlayerMixin.java:
```diff
@@ -50,7 +54,14 @@ private void postTick(CallbackInfo callbackInfo) {
 		}
 
-		var idleTime = Util.getMillis() - this.getLastActionTime();
+    // update lastMotileTime if it is older than a stale lastActionTime
+    var lastActionTime = this.getLastActionTime();
+    if (lastActionTime < actionTimeTracker) {
+      lastMotileTime = actionTimeTracker;
+      actionTimeTracker = lastActionTime;
+    }
```
I think we might not need to track the last action time in order to update `lastMotileTime`, as `lastMotileTime` already updates whenever action is taken (`setSessile(false)` is called in `resetLastActionTime`)?

donington replied:
I believe you're right, I'll double check that when I get a chance.

donington commented:
I cleaned up the `lastMotileTime` code, you were right the solution was very simple thanks to the way it updates already.  I did a final cleanup pass as well since I just can't help myself from sneaking spaces in by accident.

bytzo committed:
Cleanup pull request #4

bytzo committed:
Update copyright owners and contributors

bytzo commented:
This looks good to me! I haven't found any issues in testing so far, so I'll merge this now. Thanks so much for the contribution!

donington commented:
Happy to help improve your already good mod :)
